dev-infrastructure/scripts/istio.sh

```diff
@@ -0,0 +1,77 @@
+#!/bin/bash
+
+set -euxo pipefail
+
+echo "********** ISTIO Upgrade Started **************"
+# Followed this guide for istio upgrade https://learn.microsoft.com/en-us/azure/aks/istio-upgrade
+# To upgrade or rollback, change the targetVersion to the desire version, and version to the current version.
+if [[ -z "$TARGET_VERSION" ]]; then
+    echo "Istio is using Target Version. Exiting script."
+    exit 0
+fi
+
+echo "********** Download istioctl**************"
+ISTIO_URL="${ISTIOCTL_URL}/${ISTIOCTL_VERSION}/istio-${ISTIOCTL_VERSION}-linux-amd64.tar.gz"
+SHA256_URL="${ISTIOCTL_URL}/${ISTIOCTL_VERSION}/istio-${ISTIOCTL_VERSION}-linux-amd64.tar.gz.sha256"
+# Download the Istioctl binary
+wget "$ISTIO_URL" -O istio-"${ISTIOCTL_VERSION}"-linux-amd64.tar.gz
+
+# Download the SHA-256 checksum file
+wget "$SHA256_URL" -O istio-"${ISTIOCTL_VERSION}"-linux-amd64.tar.gz.sha256
+
+# Verify the downloaded file
+sha256sum -c istio-"${ISTIOCTL_VERSION}"-linux-amd64.tar.gz.sha256
+
+# Check the result of the verification
+if sha256sum -c istio-"${ISTIOCTL_VERSION}"-linux-amd64.tar.gz.sha256; then
+    echo "Verification successful: The file is intact."
+else
+    echo "Verification failed: The file is corrupted."
+    exit 1
+fi
+
+tar -xzf istio-"${ISTIOCTL_VERSION}"-linux-amd64.tar.gz
+cd istio-"${ISTIOCTL_VERSION}"
+export PATH=$PWD/bin:$PATH
+echo "=========================================================================="
+
+NEWVERSION="$TARGET_VERSION"
+echo "********** Istio UpGrade Started with version ${NEWVERSION} **************"
+
+istioctl tag set "$TAG" --revision "${NEWVERSION}" --istioNamespace aks-istio-system --overwrite
+# Get the namespaces with the label istio.io/rev=$TAG
+namespaces=$(kubectl get namespaces --selector=istio.io/rev="$TAG" -o jsonpath='{.items[*].metadata.name}' | xargs -n1 echo)
+
+for ns in $namespaces; do
+    pods=$(kubectl get pods -n "$ns" -o jsonpath='{.items[*].metadata.name}'| xargs -n1 echo)
```

Comment:
Instead of making many `kubectl` client calls to the server, you can simply record the data and operate over it:

```sh
for namespace in $( kubectl get namespaces --selector=istio.io/rev="$TAG" -o jsonpath='{.items[*].metadata.name}' ); do
    pods="$( kubectl get pods --namespace "${namespace}" -o json )"
    for pod in "$( jq <<<"${pods}" --raw-output --arg NEWVERSION "${NEWVERSION}" '.items[] | select(.metadata.annotations.sidecar\.istio\.io/status | fromjson.revision != $NEWVERSION) | .metadata.name' )"; do
        owner_kind="$( jq <<"${pods}" --raw-output --arg NAME "${pod}" '.items[] | select(.metadata.name == $NAME) | .metadata.ownerReferences[0].kind' )"
        # etc
    done
done
```

Reply:
I may have messed up the shell script, they are very hard to write correctly without a REPL. Let's pair on this one. Saving the output of `get pods -o json` in a file also lets your script run super duper fast when you are testing this locally, as listing all pods might take some time on the server.